docs/manual/scala/guide/services/code/ServiceImplementation.scala

```diff
@@ -0,0 +1,204 @@
+
+    override def tick(intervalMs: Int) = ServiceCall { tickMessage =>
+      Future.successful(Source.tick(
+        intervalMs.milliseconds, intervalMs.milliseconds, tickMessage
+      ))
+    }
+    //#tick-service-call
+  }
+}
+
+package hellostream {
+
+  import com.lightbend.lagom.scaladsl.api.ServiceCall
+  import scala.concurrent.Future
+
+  class HelloStreamImpl extends HelloStream {
+    //#hello-service-call
+    override def sayHello = ServiceCall { names =>
+      Future.successful(names.map(name => s"Hello $name!"))
+    }
+    //#hello-service-call
+  }
+}
+
+package serverservicecall {
+
+  import scala.concurrent.Future
+
+  class UsingServerServiceCall extends helloservice.HelloService {
+    //#server-service-call
+    import com.lightbend.lagom.scaladsl.server.ServerServiceCall
+    import com.lightbend.lagom.scaladsl.api.transport.ResponseHeader
+
+    override def sayHello = ServerServiceCall { (requestHeader, name) =>
+      val user = requestHeader.principal
+        .map(_.getName).getOrElse("No one")
+      val response = s"$user wants to say hello to $name"
+
+      val responseHeader = ResponseHeader.Ok
+        .withHeader("Server", "Hello service")
+
+      Future.successful((responseHeader, response))
+    }
+    //#server-service-call
+  }
+}
+
+package servicecallcomposition {
+
+  import scala.concurrent.{ExecutionContext, Future}
+  import com.lightbend.lagom.scaladsl.server.ServerServiceCall
+
+  object Logging {
+    //#logging-service-call
+    def logged[Request, Response](serviceCall: ServerServiceCall[Request, Response]) =
+      ServerServiceCall.compose { requestHeader =>
+        println(s"Received ${requestHeader.method} ${requestHeader.uri}")
+        serviceCall
+      }
+    //#logging-service-call
+  }
+
+  class LoggedHelloService extends helloservice.HelloService {
+    import Logging.logged
+
+    //#logged-hello-service
+    override def sayHello = logged(ServerServiceCall { name =>
+      Future.successful(s"Hello $name!")
+    })
+    //#logged-hello-service
+  }
+
+  trait User
+
+  //#user-storage
+  trait UserStorage {
+    def lookupUser(username: String): Future[Option[User]]
+  }
+  //#user-storage
+
+  object Authentication {
+    val userStorage: UserStorage = _
+    implicit val ec: ExecutionContext = _
+
+    //#auth-service-call
+    import com.lightbend.lagom.scaladsl.api.transport.Forbidden
+
+    def authenticated[Request, Response](
+      serviceCall: User => ServerServiceCall[Request, Response]
+    ) = ServerServiceCall.composeAsync { requestHeader =>
+
+      // First lookup user
+      val userLookup = requestHeader.principal
+        .map(principal => userStorage.lookupUser(principal.getName))
+        .getOrElse(Future.successful(None))
+
+      // Then, if it exists, apply it to the service call
+      userLookup.map {
+        case Some(user) => serviceCall(user)
+        case None => throw Forbidden("User must be authenticated to access this service call")
+      }
```

Comment:
The problem with 403 is this little requirement from the HTTP spec:

> The response MUST include a WWW-Authenticate header field (section 14.47) containing a challenge applicable to the requested resource.

Since we don't support any of the WWW authentication schemes, what would we put in that header?  This is why in practice, 401 response are rarely used, since no one actually uses the semantic HTTP authentication practices.  The statement on the 403 status code:

> Authorization will not help and the request SHOULD NOT be repeated.

Is taken to mean "Passing an HTTP Authorization header will not help", and that's true, both in this case, and generally in the web at large.